rls/src/main/java/io/grpc/rls/internal/CachingRlsLbClient.java

```diff
@@ -0,0 +1,962 @@
+    timeProvider = checkNotNull(builder.timeProvider, "timeProvider");
+    throttler = checkNotNull(builder.throttler, "throttler");
+    linkedHashLruCache =
+        new RlsAsyncLruCache(
+            rlsConfig.getCacheSizeBytes(),
+            builder.evictionListener,
+            scheduledExecutorService,
+            timeProvider);
+    RlsRequestFactory requestFactory = new RlsRequestFactory(lbPolicyConfig.getRouteLookupConfig());
+    rlsPicker = new RlsPicker(requestFactory);
+    rlsChannel = helper.createResolvingOobChannel(rlsConfig.getLookupService());
+    helper.updateBalancingState(ConnectivityState.CONNECTING, rlsPicker);
+    rlsStub = RouteLookupServiceGrpc.newStub(rlsChannel);
+    childLbResolvedAddressFactory =
+        checkNotNull(builder.resolvedAddressFactory, "resolvedAddressFactory");
+    backoffProvider = builder.backoffProvider;
+    ChildLoadBalancerHelperProvider childLbHelperProvider =
+        new ChildLoadBalancerHelperProvider(helper, new SubchannelStateManagerImpl(), rlsPicker);
+    if (rlsConfig.getRequestProcessingStrategy()
+        == RequestProcessingStrategy.SYNC_LOOKUP_CLIENT_SEES_ERROR) {
+      refCountedChildPolicyWrapperFactory =
+          new RefCountedChildPolicyWrapperFactory(
+              childLbHelperProvider, new BackoffRefreshListener());
+    } else {
+      refCountedChildPolicyWrapperFactory =
+          new RefCountedChildPolicyWrapperFactory(childLbHelperProvider, null);
+    }
+  }
+
+  @CheckReturnValue
+  private ListenableFuture<RouteLookupResponse> asyncRlsCall(RouteLookupRequest request) {
+    final SettableFuture<RouteLookupResponse> response = SettableFuture.create();
+    if (throttler.shouldThrottle()) {
+      response.setException(new ThrottledException());
+      return response;
+    }
+    rlsStub.withDeadlineAfter(callTimeoutNanos, TimeUnit.NANOSECONDS)
+        .routeLookup(
+            REQUEST_CONVERTER.convert(request),
+            new StreamObserver<io.grpc.lookup.v1.RouteLookupResponse>() {
+              @Override
+              public void onNext(io.grpc.lookup.v1.RouteLookupResponse value) {
+                response.set(RESPONSE_CONVERTER.reverse().convert(value));
+              }
+
+              @Override
+              public void onError(Throwable t) {
+                response.setException(t);
+                throttler.registerBackendResponse(false);
+              }
+
+              @Override
+              public void onCompleted() {
+                throttler.registerBackendResponse(true);
+              }
+            });
+    return response;
+  }
+
+  /**
+   * Returns async response of the {@code request}. The returned value can be in 3 different states;
+   * cached, pending and backed-off due to error. The result remains same even if the status is
+   * changed after the return.
+   */
+  @CheckReturnValue
+  public final CachedRouteLookupResponse get(final RouteLookupRequest request) {
+    synchronized (lock) {
```

Comment:
Can you and `synchronizationContext.throwIfNotInThisSynchronizationContext()` here?